cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/operations/InMemoryOperationRunner.java

```diff
@@ -0,0 +1,116 @@
+import io.cdap.cdap.proto.operationrun.OperationType;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link OperationRunner} to run an operation in the same service.
+ */
+public class InMemoryOperationRunner<T> extends AbstractOperationRunner<T> {
+
+  private final OperationStatePublisher statePublisher;
+  private LongRunningOperation<T> currentOperation;
+  private final Lock lock;
+  private final ExecutorService executor;
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryOperationRunner.class);
+
+  /**
+   * Default constructor.
+   */
+  @Inject
+  public InMemoryOperationRunner(Map<OperationType, LongRunningOperation> operations,
+      OperationStatePublisher statePublisher) {
+    super(operations);
+    this.statePublisher = statePublisher;
+    this.lock = new ReentrantLock();
+    this.executor = Executors.newSingleThreadExecutor();
+  }
+
+  // TODO(samik) It is currently rudimentary implementation to unblock the state management
+  //  Handle states, publish failures and heartbeat messages
+  @Override
+  public void run(LongRunningOperationRequest<T> request)
+      throws ClassNotFoundException, OperationTypeNotSupportedException {
+    this.lock.lock();
+    try {
+      if (currentOperation != null) {
+        throw new IllegalStateException("Operation already running");
+      }
+      this.currentOperation = getOperation(request);
+      executor.execute(() -> {
+        this.runInternal(request);
+      });
+      statePublisher.publishRunning();
+    } finally {
+      this.lock.unlock();
+    }
+  }
+
+  private void runInternal(LongRunningOperationRequest<T> request) {
+    try {
+      OperationError error = currentOperation.run(request.getOperationRequest(),
+          this::sendMetaUpdateMessage);
+      if (error != null) {
+        statePublisher.publishFailed(error);
+      } else {
+        statePublisher.publishSuccess();
+      }
+    } catch (Exception e) {
+      statePublisher.publishFailed(
+          getOperationErrorFromException(request, e));
+    }
+  }
+
+  @Override
+  public void stop() {
+    this.lock.lock();
+    if (this.currentOperation == null) {
+      throw new IllegalStateException("No operation is running");
+    }
+    this.executor.shutdownNow();
+    this.lock.unlock();
+    statePublisher.publishStopped();
+  }
+
+  private void sendMetaUpdateMessage(OperationMeta meta) {
+    statePublisher.publishMetaUpdate(meta);
+  }
+
+  private OperationError getOperationErrorFromException(LongRunningOperationRequest<T> request,
+      Exception e) {
+    OperationRunId runId = request.getRunId();
+    LOG.error("Operation {} of namespace {} failed: {}", runId.getRun(), runId.getParent(),
```

Comment:
When logging exception, always have the exception as the last parameter so that the stacktrace is getting logged.

Reply:
Is this an error log? Seems more an info or debug to me.

Reply:
done